### DISCUSSION TITLE
Is there a way to test ref structs?

### DISCUSSION THREAD

**Discussion by yaakov-h** [thomhurst/TUnit#839] [Q&A]
I'm trying out TUnit in a brand new projects and have written a `ref struct` that I want to assert the properties of.

Assertions are async and must be awaited, so I get this compiler error:

> Class1.cs(28,27): error CS4007: Instance of type 'MyRefStruct' cannot be preserved across 'await' or 'yield' boundary.

Is there a way to test these with TUnit or does the async nature of the assertions framework prevent this?

**Comment by thomhurst** [comment]
This should be possible in the next C# language version.

I believe you can trial it out already by installing the .NET 9 SDK and then setting `<LangVersion>preview</LangVersion>`

You won't be able to pass your ref struct directly into the Assert.That() but you should be able to pass a property into it fine.

**Reply by yaakov-h** [reply]
I'm on .NET 9 RC2 already. I've now tried setting LangVersion=preview but I get the same compiler error.

Looking over the C# 13 design closer, the `await` and the `ref struct` have to be in different block scopes, e.g.:

```cs
string x;

{
    x = GetRefStruct().Property;
}

await Assert.That(x).IsEqualTo("whatever");
```

However I cannot come up with a decent solution for my case which involves asserting the `ref struct` results of a `ref struct` enumerator...

**Reply by thomhurst** [reply]
What does your code look like?

This compiles fine for me on the preview language version:
https://github.com/thomhurst/TUnit/blob/main/TUnit.Assertions.UnitTests/RefStructEqualToAssertionTests.cs
